### PR TITLE
Linter: SASSLintBear

### PR DESCRIPTION
**Pull request by sopelj**
Added a sass-lint Bear for linting SASS/SCSS files

Closes https://github.com/coala/coala-bears/issues/982

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [X ] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [X ] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

File: bears/scss/SASSLintBear.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        return args
+
+    def process_output(self, output, filename, file):
+        if not file or not output:
+            return
+
+        try:
+            output = json.loads(output)
+        except ValueError:
+            self.warn('Failed to pass output of sass-lint')
```
Comment (blazeu) [inline bears/scss/SASSLintBear.py]: and what is the output looks like? the user needs to see the errors (print it)

**Comment by blazeu** [comment]
I think `bears` are preferable in the shortlog tag instead of `Linter`

Also you don't need to explain what you're doing in commit body if you use `Add SASSLintBear` in shortlog ;)

File: bears/scss/SASSLintBear.py (hunk 1 above)
Comment (sopelj) [reply bears/scss/SASSLintBear.py]: Good point

**Comment by sopelj** [comment]
I tried to test everything, but it says there is only 43% coverage.
I looked at other bears and they all seem to test about the same thing, but they pass. I'm clearly missing something...

File: bears/scss/SASSLintBear.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+from coalib.results.RESULT_SEVERITY import RESULT_SEVERITY
+from coalib.results.Result import Result
+
+
+@linter(executable='sass-lint')
+class SASSLintBear:
+    """
+    Check SCSS/SASS code to keep it clean and readable.
+
+    More information is available at <https://github.com/sasstools/sass-lint>.
```
[the remark this answers is not in the document]
Comment (sopelj) [reply bears/scss/SASSLintBear.py]: Good point

File: bears/scss/SASSLintBear.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    @staticmethod
+    def create_arguments(filename, file, config_file, sasslint_config: str=''):
+        """
+        :param sasslint_config: The location of the sass-lint config file.
+        """
+        args = tuple()
+
+        if os.path.splitext(filename)[1] == '.scss':
+            args += ('--syntax=scss',)
+
```
Comment (RaiVaibhav) [inline bears/scss/SASSLintBear.py]: please remove space

File: bears/scss/SASSLintBear.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        if sasslint_config:
+            args += ('--config=' + sasslint_config,)
+
+        args += ('--format=json', '--verbose', '--no-exit', filename)
+
```
Comment (RaiVaibhav) [inline bears/scss/SASSLintBear.py]: line 40,42, 44 please remove space accordingly for more.

File: tests/scss/SASSLintBearTest.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    padding: 0
+
+  li
+    display: inline-block
+
+  a
+    display: block
+    padding: 6px 12px
+    text-decoration: none
+
```
Comment (RaiVaibhav) [inline tests/scss/SASSLintBearTest.py]: same here extra space please remove it and for others below

File: tests/scss/SASSLintBearTest.py [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+  li { display: inline-block; }
+
+  a {
+    display: block;
+    padding: 6px 12px;
+    text-decoration: none;
+}
+"""
+
+
```
Comment (RaiVaibhav) [inline tests/scss/SASSLintBearTest.py]: 2 line space make it 1 please,even it will work without any space

File: tests/scss/SASSLintBearTest.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+test_dir = os.path.join(os.path.dirname(__file__), 'test_files')
+
+
+SASSLintBearTest = verify_local_bear(
+    SASSLintBear,
+    valid_files=(good_sass_file,),
+    invalid_files=(bad_sass_file,),
+    tempfile_kwargs={'suffix': '.sass'}
+)
```
Comment (RaiVaibhav) [inline tests/scss/SASSLintBearTest.py]: please move( `)`) it to on above line. :smile:

Comment (RaiVaibhav) [reply tests/scss/SASSLintBearTest.py]: same for others.

File: bears/scss/SASSLintBear.py (hunk 4 above)
Comment (sopelj) [reply bears/scss/SASSLintBear.py]: Looks a bit squished, but ok.

File: tests/scss/SASSLintBearTest.py (hunk 7 above)
Comment (sopelj) [reply tests/scss/SASSLintBearTest.py]: ok

**Comment by yash-nisar** [comment]
We generally start working on an issue after being assigned to it to avoid conflicts. So, kindly take care of this in future 😄

**Comment by sopelj** [comment]
Oh, that's a very good point. I hadn't thought to modify the package.json. That could definitely explain the tests not passing on the test server. I added it and I will see if it passes now. Yes, it does pass locally.

Sorry, I'm not familiar with the protocol. Should I have requested assignment somehow?

**Comment by yash-nisar** [comment]
> Sorry, I'm not familiar with the protocol. Should I have requested assignment somehow?

Yes, but it's okay if you were not aware of it. ;)

**Comment by sopelj** [comment]
Oh ok, thanks. By which means should I make a request like this in future?
The `ci/circleci` check fails, but it seems to be related to another module as it says :

> Warning, treated as error:
> /home/ubuntu/coala-bears/docs/API/bears.vcs.rst:document isn't included in any toctree

**Comment by sopelj** [comment]
I updated my branch with coala-bear master and everything seems ok now. Although, the ci-circle fails, but that seems to caused by #2309

**Comment by jayvdb** [comment]
@gitmate-bot rebase

**Comment by sopelj** [comment]
Excllent :smile:  Now that #2309 is fixed it circle is passing.

**Comment by sopelj** [comment]
Sorry, wrong button.

File: bears/scss/SASSLintBear.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    LICENSE = 'AGPL-3.0'
+    CAN_DETECT = {'Syntax', 'Formatting'}
+    SEE_MORE = 'https://github.com/sasstools/sass-lint'
+
+    severity_map = {2: RESULT_SEVERITY.MAJOR,
+                    1: RESULT_SEVERITY.NORMAL,
+                    0: RESULT_SEVERITY.INFO}
+
+    @staticmethod
+    def create_arguments(filename, file, config_file, sasslint_config: str=''):
```
[the remark this answers is not in the document]
Comment (sopelj) [reply bears/scss/SASSLintBear.py]: Ok, it has been changed

File: package.json [hunk 9]
```diff
@@ -24,6 +24,7 @@
     "remark-cli": "~2",
     "remark-lint": "~5",
     "remark-validate-links": "~5",
+    "sass-lint": "~1.12.1",
```
Comment (jayvdb) [inline package.json]: there is an update to the generator, which also put the requirements in `bear-requirements.yaml`. Se travis log.

**Comment by jayvdb** [comment]
ack https://github.com/coala/coala-bears/pull/2122/commits/7b219b7881cab7aa3f3fa4b43fd2b4558df3f53a

**Comment by jayvdb** [comment]
@gitmate-bot rebase

**Comment by jayvdb** [comment]
@gitmate-bot rebase

**Comment by jayvdb** [comment]
@gitmate-bot ff

**Comment by jayvdb** [comment]
sorry I had forgotten about this one ;-)

**Pull request merged**